### PR TITLE
UIDEXP-22: Implement DataFetcher component for making API requests upon interval

### PR DESCRIPTION
**Pull request by VictorSoroka1**
## Purpose

Implement `DataFetcher` component for making API requests upon interval in the scope of the [UIDEXP-22](https://issues.folio.org/browse/UIDEXP-22).

### CODE REVIEW

File: lib/DataFetcher/tests/DataFetcher-test.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        });
+      }
+
+      Object.assign(fetcherData.resources, { logs: { records: [[{ id: 'log1Id' }]] } });
+    }, 300);
+  }
+
+  await mount(
+    <DataFetcher {...fetcherData}>
+      <FetcherContext.Consumer>{value => { fetcherData.fetcherContextValue = value; }}</FetcherContext.Consumer>
```
Comment (Yuliia-Dovhal) [inline lib/DataFetcher/tests/DataFetcher-test.js]:
```suggestion
      <FetcherContext.Consumer>
          {value => { fetcherData.fetcherContextValue = value; }}
      </FetcherContext.Consumer>
```

File: lib/DataFetcher/DataFetcher.js [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+          hasLoaded: true,
+        }));
+      }
+    }
+  };
+
+  useEffect(() => resetFetcher, []);
+
+  useEffect(() => {
+    function updateResourcesData(initial) {
```
Comment (YevheniiMaltsev) [inline lib/DataFetcher/DataFetcher.js]: In some cases (hasLoaded, isEmpty) we use "has" or "is" to specify the boolean variable. But in this case "has" or " is" is not used. I think we should select and use one approach for consistency.

File: lib/DataFetcher/DataFetcher.js [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+      if (!initial && timeoutId.current === null) return;
+
+      resetFetcher();
+
+      timeoutId.current = setTimeout(async () => {
+        await fetchResources(initial);
+
+        updateResourcesData();
```
Comment (YevheniiMaltsev) [inline lib/DataFetcher/DataFetcher.js]: In the line 82 the boolean value "true" is used, but in this case the value "undefined" will be passed. Let's setup a default value  "false" (initial = "false") or pass "false" to keep types consistency.

File: lib/DataFetcher/DataFetcher.js [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+  const resetFetcher = () => {
+    clearTimeout(timeoutId.current);
+    timeoutId.current = null;
+  };
+
+  const fetchResources = async initial => {
+    const resourcesPromises = Object
+      .values(mutator)
+      .reduce((res, resourceMutator) => res.concat(fetchResource(resourceMutator)), []);
```
[the remark this answers is not in the document]
Comment (VictorSoroka1) [reply lib/DataFetcher/DataFetcher.js]: `resource` would not pass the sonar check as there is such name at higher scope. `res` stands for `result`. Can name it like that.

[the remark this answers is not in the document]
Comment (YevheniiMaltsev) [reply lib/DataFetcher/DataFetcher.js]: Okay, let's name it "result" if it stands for result

File: lib/DataFetcher/DataFetcher.js [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+      timeoutId.current = setTimeout(async () => {
+        await fetchResources(initial);
+
+        updateResourcesData();
+      }, initial ? 0 : updateInterval);
+    }
+
+    updateResourcesData(true);
+  }, [updateInterval]); // eslint-disable-line react-hooks/exhaustive-deps
```
[the remark this answers is not in the document]
Comment (VictorSoroka1) [reply lib/DataFetcher/DataFetcher.js]: I such a case I don't think about such. If I add the entities which lint rule requires, the app would not work correctly as the arguments passed to that array are checked based on the shallow comparison. If I provide `mutator` there, then I need to do a deep comparison which does not make sense. I treat this rule as the helper to identify what can be improved. Following this rule is not always required.

File: lib/DataFetcher/DataFetcher.js (hunk 2 above)
Comment (VictorSoroka1) [reply lib/DataFetcher/DataFetcher.js]: Sounds good.

File: lib/DataFetcher/DataFetcher.js (hunk 3 above)
Comment (VictorSoroka1) [reply lib/DataFetcher/DataFetcher.js]: Sounds good. Thank for good comments @YevheniiMaltsev

**Pull request merged**